aiounifi/interfaces/traffic_rules.py

```diff
@@ -0,0 +1,31 @@
+"""WLANs as part of a UniFi network."""
+
+from typing import Any
+
+from ..models.traffic_rule import TrafficRule, TrafficRuleListRequest, TrafficRuleEnableRequest
+from .api_handlers import APIHandler
+
+
+class TrafficRules(APIHandler[TrafficRule]):
+    """Represents TrafficRules configurations."""
+
+    obj_id_key = "_id"
+    path = "/trafficrules"
+    item_cls = TrafficRule
+    api_request = TrafficRuleListRequest.create()
+
+    async def enable(self, traffic_rule: TrafficRule) -> list[dict[str, Any]]:
+        """Enable traffic rule defined in controller."""
+        tr = self.controller.traffic_rules.get(traffic_rule.id)
+        traffic_rule_dict = tr.raw
+        return await self.controller.request(
+            TrafficRuleEnableRequest.create(traffic_rule_dict, enable=True)
+        ) and await self.controller.traffic_rules.update()
```

Comment:
Well, I'm also convinced that is not the right place to trigger an update (I already removed this locally). I just can't seem to grasp how the state of a traffic rule will be updated, so that the state in the controller reflects the actual state of the traffic rule. Or is this not necessary?

I inspected the messages coming from the Unifi-controller (websocket) and I don't see any messages emitted when changing the state of the traffic rule.

Reply:
Do the controller only do a get on the endpoint when changing configuration?

Reply:
The controller does a PUT on the endpoint. The response is the a JSON with the changed configuration (without meta/data-keys, just the config of the traffic rule). I have been monitoring the traffic for a while (+/- 30 minutes, also on the websocket) and I have not seen any other traffic passing concerning the PUT on the endpoint when changing the state.

Reply:
> The controller does a PUT on the endpoint. The response is the a JSON with the changed configuration (without meta/data-keys, just the config of the traffic rule). I have been monitoring the traffic for a while (+/- 30 minutes, also on the websocket) and I have not seen any other traffic passing concerning the PUT on the endpoint when changing the state.

Maybe we should pass the response back into update in the handler if you get the updated response? On the TrafficRule handler class. Consolidate enable disable methods to one to avoid the duplication and I think that could work well.




> I assume there is a slight problem in [api_handlers.py](https://github.com/Kane610/aiounifi/blob/master/aiounifi/interfaces/api_handlers.py#L99) the trafficrules-api in this line:
> 
> `self.process_raw(raw.get("data", []))`
> 
> This assumes that raw is a dict. But since the answer of the trafficrules-api is an array (of JSON-objects), the orjson.loads-function converts this [array to a list](https://docs.python.org/3/library/json.html#json-to-py-table) - I believe they also follow this conversion table.
> 
> Maybe we need to see how this affects the library and how I can best work around this.

Im considering moving initial response data parsing to the response object, could host both error parsing and preparing response for the handler.

Reply:
- for the data parsing of the response object, I've made a suggestion by adding a 'prepare_data' method to the ApiRequest-class (not yet completely happy with this, but it is a start). And added a call in the update()-method of the ApiHandler
- I have consolidated the enable/disable-methods (they still exist, but they call a new method 'toggle'), so code duplication is avoided. Still have to figure out a way to use the response to make the update. I'm now calling the update()-method of the handler in toggle, but using the response would be way more efficient.